lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,379 @@
+        return false;
+    }
+}
+
+void VisSharedMemWriter::reset_memory(uint32_t time_ind) {
+
+    // resets all memory at time_ind to 0s
+
+    uint8_t *buf_write_pos = buf_addr + (time_ind * nfreq * frame_size);
+    int64_t *access_record_write_pos = access_record_addr + (time_ind * nfreq);
+
+    DEBUG("Resetting access_record memory at position time_ind: {}\n", time_ind);
+
+    // notify that the entire time_ind is invalid, by setting time_ind in the access record to invalid
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}", _sem_name);
+        return;
+    }
+    int64_t invalid_vector[nfreq];
+    std::fill_n(invalid_vector, nfreq, invalid);
+    memcpy(access_record_write_pos, invalid_vector, nfreq * sizeof(invalid_vector[0]));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to post semaphore {}", _sem_name);
+        return;
+    }
+
+    INFO("Resetting ring buffer memory at position time_ind: {}\n", time_ind);
+    // set the full time_ind to 0 in the ring buffer
+    std::vector<char> zeros(nfreq * frame_size, 0);
+    char* tmp = zeros.data();
+    memcpy(buf_write_pos, &tmp, sizeof(tmp));
+
+    INFO("Memory reset\n");
+}
+
+void VisSharedMemWriter::write_to_memory(const visFrameView& frame, uint32_t time_ind, uint32_t freq_ind) {
+    // write frame to ring buffer at time_ind and freq_ind
+
+    uint8_t *buf_write_pos = buf_addr + ((time_ind * nfreq + freq_ind) * frame_size);
+    int64_t *access_record_write_pos = access_record_addr + (time_ind * nfreq + freq_ind);
+
+    DEBUG("Writing ringbuffer to time_ind {} and freq_ind {}\n", time_ind, freq_ind);
+
+    // notify that time_ind and freq_ind are being written to, by setting that
+    // location to invalid in the access record
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {}. Dropping frame.\n", _sem_name);
+        return;
+    }
+    memcpy(access_record_write_pos, &invalid, sizeof(invalid));
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+
+    // first write the metadata, then the data, then the valid byte
+    // add valid_size amount of padding
+    memcpy(buf_write_pos + valid_size, frame.metadata(), metadata_size);
+    memcpy(buf_write_pos + metadata_size + valid_size, frame.data(), data_size);
+    memcpy(buf_write_pos, &valid, sizeof(valid));
+
+    // Document the fpga sequence counter for that frame in the access record
+    uint64_t fpga_seq = frame.metadata()->fpga_seq_start;
+
+
+    if (wait_for_semaphore() == false) {
+        ERROR("Failed to acquire semaphore {} before metadata was updated.\n", _sem_name);
+        return;
+    }
+
+    DEBUG("Writing fpga_seq {} to time index {}\n", fpga_seq, time_ind);
+    memcpy(access_record_write_pos, &fpga_seq, sizeof(fpga_seq));
+
+    // update num_writes
+    num_writes++;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
```

Comment:
Again. Just assign directly.

Definitely looks like you've been bought off by the National Memcpy Council.